Test completed: AWS CDK Deployment for Scalable and Resilient React Application

pablomagro opened the pull request:
The repository includes two new directories: `test-five` and `.github/workflows,` each of which contains the following:

* GitHub Actions for testing, building, deployment, and vulnerability checks.
* An `Infrastructure` folder, which houses the project's CDK infrastructure.
* A `Frontend` folder, containing a basic React application.

pablomagro committed:
Create Test Five project folder.

Create README.md file.

pablomagro committed:
feat: Create Test Five project structure.

- Add React TypeScript from the provided template.
- Docker and compose files for local testing.

pablomagro committed:
feat: Add CDK with the React Infrastructure.

- Add deploy.sh helper for local tests.
- Add unit tests, not working yet.
- Override parent .gitignore rule for the Lib folder.

pablomagro committed:
doc: Add deployment images.

Update the `README.md` file with details about the infrastructure and the outcome of the application deployment.

pablomagro committed:
feat: Adds GitHub Action app deployment pipeline.

- It involves the building and deploying of the
connected infrastructure within the `deploy.yml`
GitHub Action workflow.

- Note that unit tests and end-to-end (E2E) tests
are not included at this time.

pablomagro committed:
feat: Added GitHub Action Unit Tests.

In this update, it includes both frontend and infrastructure unit tests. Additionally,  it's added their dependencies to the build and deployment pipeline.

pablomagro committed:
fix: Update App.test.tsx unit test.

It updates the test with the right text to search.

pablomagro committed:
chore: Update Infrastructure Node project file.

- Set the minimum `Node.js` version to 20.
- Add the `test-cov` script for unit tests and coverage.

pablomagro committed:
fix: Update Infrastructure folder path.

pablomagro committed:
fix: Update Infrastructure folder path.

pablomagro committed:
chore: Set to 20 the project node version.

pablomagro committed:
chore: Add Trivy Scanner for vulnerabilities

Trivy is an open-source vulnerability scanner designed to detect security issues in container images and filesystems.

pablomagro committed:
doc: Add more comments to infrastructure-stack.ts file.

pablomagro committed:
chore: Add E2E tests to GitHub action deploy pipeline.

The Cypress library has been added to perform frontend behavioral tests.

pablomagro committed:
chore: Add E2E tests to GH action deploy pipeline.

The Cypress library has been added to perform frontend behavioral tests.

pablomagro committed:
fix: Add missed hidden files.

pablomagro committed:
fix: Delete file from wrong folder.

pablomagro committed:
fix: Delete file from wrong folder.

pablomagro committed:
fix: Delete file from wrong folder.

pablomagro committed:
fix: Add missed hidden files.

pablomagro committed:
fix: Add missed hidden files.

pablomagro committed:
Refactor: Clean up code - remove console.log statement.

pablomagro committed:
feat: Add SonarQube GitHub action for code checks.

Add this action to the code scanning process to detect bugs, vulnerabilities, and code smells once unit tests have been passed.

pablomagro committed:
feat: Adds an alarm to monitor S3 and CloudFront.

Alarms crated:

* `S3`: Sets up an alarm that triggers if there are more than 100 errors in a 5-minute period for two consecutive periods.
* `CloudFront`:  Alarm that triggers if the average 4xx error rate exceeds 1% in a 5-minute period for two consecutive periods.

pablomagro committed:
chore: Add the parameter `SnsRecipient`.

The SnsRecipient parameter is now required to deploy the application's CDK infrastructure.

pablomagro committed:
refactor: Add required SNSRecipient parameter.

The `SnsRecipient` parameter is now required to deploy the application's CDK infrastructure.

pablomagro committed:
fix: Update wrong access to STACK_ENVIRONMENT env variable.

pablomagro committed:
feat: Create a Dashboard for managing alarms.

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+name: Test, Build, and Deploy - CI/CD
+
+on:
+  workflow_dispatch:
```
good use of these triggers

pablomagro replied:
Oops, it shouldn't be there. I only used it for testing.

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+          npm install -g ts-node
+          npm install -g --unsafe-perm=true esbuild
+
+      - name: Infrastructure Test
```
great to see unit tests

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+        with:
+          # Disabling shallow clone is recommended for improving relevancy of reporting.
+          fetch-depth: 0
+      - name: SonarQube Scan
```
very cool

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+          login: ${{ secrets.SONARQUBE_TOKEN }}
+
+  build-and-deploy:
+    needs: [unit-test-frontend, unit-test-infrastructure]
```
I like this chaining

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@master
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
```
glad to see use of secrets rather than adding creds into a public repo 😄

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+      - name: Deploy to ${{ env.STACK_ENVIRONMENT }} AWS
+        run: |
+          cd infrastructure
+          cdk deploy TestFiveStack --parameters Environment=${{ env.STACK_ENVIRONMENT }} --parameters SnsRecipient=${{ secrets.SNS_RECIPIENT }} --require-approval never --progress=events --no-cli-pager
```
This is what we wanted to see, great. and with SNS notification + 1

danwfeeney commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,157 @@
+
+      # Install NPM dependencies, cache them correctly
+      # and run all Cypress tests
+      - name: Cypress run on Chrome
```
And e2e tests +1 very thorough.

pablomagro replied:
The purpose of this is to illustrate that the project must include end-to-end (E2E) tests. Additionally, compatibility testing, primarily targeting different browsers and devices, should be incorporated. [BrowserStack](https://www.browserstack.com/) is a suitable framework that can be utilized for this purpose.

danwfeeney commented on .github/workflows/repo-vulnerability.yml:
```diff
@@ -0,0 +1,40 @@
+  workflow_dispatch:
+  push:
+    branches:
+      - master
```
why only on master? wouldn't we want to protect master?

pablomagro replied:
The action is triggered every time there is a push to the main branch or for every pull request. I adopted this behavior from the [Trivy Action](https://github.com/aquasecurity/trivy-action) project, which, in my opinion, makes sense.

To address your questions, you are right, if we are using, for example, the Gitflow branching model, we should consider including the develop branch, release branches, hotfix branches, or feature branches as well. The inclusion depends on the branching model used by the team. I was thinking of the 'trunk' model based on my previous job.

danwfeeney commented on test-five/README.md:
```diff
@@ -0,0 +1,55 @@
+# AWS Services and Architecture Design.
+
+The following AWS services and architecture can be used to deploy a React application to AWS using TypeScript and CDK, considering scalability, resilience, and cost-effective:
```
Really appreciate this explanation. All makes sense too and addresses the requirements described.

danwfeeney commented on test-five/deploy.sh:
```diff
@@ -0,0 +1,27 @@
+# Script used for local tests.
+# Note that below env variables need to be informed in order to make the deployment work.
+
+cd frontend
```
can you explain why you would take this shell approach instead of npm scripts?

pablomagro replied:
I originally created this script solely for local infrastructure testing. However, I later decided to incorporate it into the project as a helpful tool. It now makes complete sense to automate the process and include it in the package.json scripts section within the infrastructure folder for use in different environments, such as e2e, testing, and production.

```json
"build:frontend": "cd ../frontend && npm run build",
"deploy:e2e": "npm run build:frontend && cdk deploy TestFiveStack --parameters Environment=e2e --parameters SnsRecipient=moshtix-infra-e2e@gmail.com --require-approval never --progress=events",
"deploy:testing": "npm run build:frontend && cdk deploy TestFiveStack --parameters Environment=testing --parameters SnsRecipient=moshtix-infra-test@gmail.com --require-approval never --progress=events",
"deploy:production": "npm run build:frontend && cd ../infrastructure && cdk deploy TestFiveStack --parameters Environment=production --parameters SnsRecipient=moshtix-infra-prod@gmail.com --require-approval never --progress=events"
```

Here's an example of how you can create the testing infrastructure:

```zsh
DOMAIN_NAME="example.com" CDK_DEFAULT_ACCOUNT=123456789 CDK_DEFAULT_REGION="ap-southeast-2" npm run deploy:testing
```

This script can also be seamlessly integrated into your CI/CD workflow, as the required environment variables are already included. Additionally, you can leverage AWS SSM to retrieve parameters if needed.rameters.

danwfeeney commented on test-five/docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+version: '3'
```
can you explain why you chose to introduce docker here?

pablomagro replied:
Docker references should not be present in this project because my solution does not utilise Docker. Initially, I intended to employ Docker for testing, deployment, and inclusion in the pipeline for deployment on Beanstalk or using Fargate. Consequently, I included Docker and Docker Compose files. However, considering the testing requirements, I decided to opt for an alternative architecture that does not rely on Docker. I forgot to delete those resources, even though it can still be used for local development.

danwfeeney commented on test-five/frontend/Dockerfile.dev:
```diff
@@ -0,0 +1,17 @@
+# Use an official Node.js runtime as the base image
```
can you explain why you chose to introduce docker here?

pablomagro replied:
I have responded to this question [here](https://github.com/moshtix/show-us-what-you-got/pull/69/files/892c3f395fc63cb15c59cc1e4e640bfc04320e5d#r1378331991). I should have removed all Docker references as it's not used in the project, serving the sole purpose of local development.

danwfeeney commented on test-five/frontend/cypress/e2e/spec.cy.ts:
```diff
@@ -0,0 +1,8 @@
+
+describe('CRA', () => {
+  it('shows learn link', function () {
+    cy.visit('http://localhost:3000')
```
is this going to work with CI/CD given it's hardcoded to localhost?

pablomagro replied:
It works within the CI/CD process as the plugin used in the workflow utilizes the secret environment [${{ secrets.FRONTEND_BASE_URL }}](https://github.com/pablomagro/show-us-what-you-got/blob/892c3f395fc63cb15c59cc1e4e640bfc04320e5d/.github/workflows/deploy.yml#L157), which is set to http://localhost:3000. This plugin installs, builds, and starts the server locally within the workflow.

Locally, it also functions when the local server is started. However, as a best practice, it should be referenced using an environment variable, as shown below:

```javascript
cy.visit(process.env.FRONTEND_BASE_URL)
```

Alternatively, you can achieve this by using a configuration library, such as config.ts, to read AWS SSM parameters for each environment. Here's an example:

```javascript
cy.visit(config.e2e.baseUrl)
```

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+    super(scope, id, props);
+
+    /**
+     * 👉 Environment Variable:
```
emoji's in code comments haha what will they think of next

pablomagro replied:
I sometimes use emojis in pull request titles, scripts, or within the source code to emphasize specific points, but whether or not I do so depends on the project and the team's preferences...

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+     * 👉 Environment Variable:
+     */
+
+    const envParameter = new CfnParameter(this, "Environment", {
```
+1 for environments. This hasn't really been applied right through CI/CD though right? Not saying you should do the work for this, just clarifying.

pablomagro replied:
Yes, it has already been integrated into the CI/CD workflow. Please review [this line of code](https://github.com/pablomagro/show-us-what-you-got/blob/892c3f395fc63cb15c59cc1e4e640bfc04320e5d/.github/workflows/deploy.yml#L126C31-L126C31).

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+    // Create an S3 bucket for the React app.
+    const reactAppBucket = new s3.Bucket(this, "ReactAppBucket", {
+      bucketName: `moshtix-fronted-${envParameter.valueAsString}`,
+      publicReadAccess: false,
```
great

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+      ],
+      viewerCertificate: cloudfront.ViewerCertificate.fromAcmCertificate(siteCertificate, {
+        aliases: [DOMAIN_NAME, `www.${DOMAIN_NAME}`],
+        securityPolicy: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
```
+1 for 1.2

pablomagro replied:
The `TLS_V1_2_2021` protocol is the recommended default protocol in AWS CloudFront, and this choice makes sense for several reasons: security, compatibility, performance, regulatory compliance, and future-proofing.

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+    );
+
+    // Sets up an alarm that triggers if there are more than 100 errors in a 5-minute period for two consecutive periods.
+    const s3BucketSizeAlarm = new cloudwatch.Alarm(this, 'HighErrorRate', {
```
great demo of this in CDK

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+    s3BucketSizeAlarm.addAlarmAction(new cw_actions.SnsAction(snsTopic));
+
+    // Alarm that triggers if the average 4xx error rate exceeds 1% in a 5-minute period for two consecutive periods.
+    const cloudWatch4xxErrorAlarm = new cloudwatch.Alarm(this, 'HighCloudFront4xxErrorRate', {
```
Again this is so great.

danwfeeney commented on test-five/infrastructure/lib/infrastructure-stack.ts:
```diff
@@ -0,0 +1,196 @@
+    cloudWatch4xxErrorAlarm.addAlarmAction(new cw_actions.SnsAction(snsTopic));
+
+    // Create a Dashboard for Monitoring and Managing Alarms
+    const dashboard = new cloudwatch.Dashboard(this, "Test Five Dashboard", {
```
+1

danwfeeney commented on test-five/infrastructure/test/infrastructure.test.ts:
```diff
@@ -0,0 +1,102 @@
+import { Template } from 'aws-cdk-lib/assertions';
+import { TestFiveStack } from '../lib/infrastructure-stack';
+
+describe.skip('TestFiveStack', () => {
```
+1

pablomagro replied:
Good catch! Somehow, I forgot to update the infrastructure unit tests as they were failing, and I wanted to complete the PR. I had forgotten to set up the environment variables.

I have completed this commit [2e3fb64](https://github.com/moshtix/show-us-what-you-got/pull/69/commits/2e3fb646bdd9a32260e07cd8103a706c3941aafe), which incorporates all the changes resulting from the code review. Additionally, I've introduced a new approach for managing Stack environment variables using the `CDK context`, which makes the entire process smoother and cleaner in terms of development, testing, and deployment.

danwfeeney reviewed (approved):
Really solid submission Pablo. Look forward to chatting on Friday. Please review and reply to comments.

pablomagro committed:
Miscellaneous Changes

As a result of the PR review, the following changes have been completed:

- Removed unnecessary Docker files.
- Fixed infrastructure unit tests by addressing missing environment variables.
- Refactored the method of obtaining deployment environment variables. It now utilizes CDK context.
- Removed an unnecessary bash deployer file.
- Added npm scripts to the infrastructure folder for application deployment.
- Updated the CI/CD deployment workflow as mentioned above.

pablomagro committed:
test: Add more infrastructure unit tests.

I had already created these tests (please note
that they are not yet 100% complete).

I will not be including additional tests/fixes
because I don't believe it is the primary objective
of these tests. However, for this kind of
development of new features, TDD and unit tests
must always be provided, as well as E2E tests.
load and performance tests will probably be
performed as well, but I think they are outside
the scope of this test.

pablomagro committed:
Miscellaneous Changes:

- Remove the `any` type for the configuration stack variables object.
- Ensure consistency in the names of deployed services according to the deployment environment.

pablomagro committed:
chore: Add deployment safety check.

pablomagro committed:
Miscellaneous Changes:

- Ensure greater consistency in the Stack name.
- Integrate an S3 origin into the CloudFormation distribution.
- Enhance the architecture design diagram with a more refined picture.

pablomagro commented:
My mistake. I resolved all the comments when I fixed or agreed with comments in the PR (we had that bad custom in my previous job). However, I personally think that if the author of the PR is always the person responsible for marking a comment as resolved, then the reviewer will always have a chance to review that fix or change...
